### PR TITLE
rgw multisite: metadata sync treats all errors as 'transient' for retry

### PR DESCRIPTION
**Pull request by cbodley**
collect_children() had a special case for EAGAIN that it treated as a 'transient' error, which set the can_adjust_marker = false to bail out of RGWMetaSyncShardCR and retry from the previous marker

but the http client doesn't return EAGAIN - rgw_http_error_to_errno() defaults to EIO - so this retry logic based on can_adjust_marker never runs. on any other error, RGWMetaSyncSingleEntryCR would not call marker_tracker->finish() to advance the sync status marker, and RGWMetaSyncShardCR would continue on with full- or incremental sync without ever attempting to retry the failed entries

a detailed comment in collect_children() describes a different strategy for handling 'permanent' errors, but that was never fully elaborated. i also don't think there's a reasonable way to differentiate between transient and permanent errors, so this treats all errors as transient to be retried

if an error really is permanent for a given metadata key, metadata sync will get stuck there and require manual intervention

Fixes: https://tracker.ceph.com/issues/39657

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/rgw/rgw_sync.cc [hunk 1]
```diff
@@ -1393,6 +1393,8 @@ class RGWCloneMetaLogCoroutine : public RGWCoroutine {
   int state_store_mdlog_entries_complete();
 };
 
+#define META_SYNC_SPAWN_WINDOW 20
```
Comment (mattbenjamin) [inline src/rgw/rgw_sync.cc:1396]: is this the new bound?

File: qa/suites/rgw/multisite/overrides.yaml [hunk 2]
```diff
@@ -14,5 +14,6 @@ overrides:
         rgw md log max shards: 4
         rgw data log num shards: 4
         rgw sync obj etag verify: true
+        rgw sync meta inject err probability: 0.1
```
Comment (mattbenjamin) [inline qa/suites/rgw/multisite/overrides.yaml:17]: this is cool

**Review by mattbenjamin — COMMENTED** [review]
I want to say lgtm--is this putatively complete?

File: qa/suites/rgw/multisite/overrides.yaml (hunk 2 above)
Comment (cbodley) [reply qa/suites/rgw/multisite/overrides.yaml:17]: yeah, the injection has been around forever, we just couldn't pass tests with it enabled

File: src/rgw/rgw_sync.cc (hunk 1 above)
Comment (cbodley) [reply src/rgw/rgw_sync.cc:1396]: yes, this is the max number of metadata entries we'll sync at a time, per mdlog shard

**Review by smanjara — APPROVED** [review]
Looks right to me

**Comment by cbodley** [comment]
https://pulpito.ceph.com/cbodley-2021-07-21_12:55:41-rgw-wip-cbodley-testing-distro-basic-smithi/

**Pull request merged**